p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,388 @@
+	for _, a := range hs.host.Peerstore().Addrs(rp) {
+		if manet.IsPublicAddr(a) && !isRelayAddress(a) {
+			forceDirectConnCtx := network.WithForceDirectDial(hs.ctx, "hole-punching")
+			dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+
+			tstart := time.Now()
+			err := hs.host.Connect(dialCtx, peer.AddrInfo{ID: rp})
+			dt := time.Since(tstart)
+			cancel()
+
+			if err != nil {
+				hs.tracer.DirectDialFailed(rp, dt, err)
+				break
+			}
+			hs.tracer.DirectDialSuccessful(rp, dt)
+			log.Debugf("direct connection to peer %s successful, no need for a hole punch", rp.Pretty())
+			return nil
+		}
+	}
+
+	// hole punch
+	for i := 0; i < maxRetries; i++ {
+		addrs, rtt, err := hs.initiateHolePunch(rp)
+		if err != nil {
+			hs.handlerError(rp, err)
+			return err
+		}
+		synTime := rtt / 2
+		log.Debugf("peer RTT is %s; starting hole punch in %s", rtt, synTime)
+
+		// wait for sync to reach the other peer and then punch a hole for it in our NAT
+		// by attempting a connect to it.
+		timer := time.NewTimer(synTime)
+		select {
+		case <-timer.C:
+			pi := peer.AddrInfo{
+				ID:    rp,
+				Addrs: addrs,
+			}
+			start := time.Now()
+			hs.tracer.StartHolePunch(rp, addrs, rtt)
+			err := hs.holePunchConnect(pi, i+1)
+			dt := time.Since(start)
+			hs.tracer.EndHolePunch(rp, dt, err)
+			if err == nil {
+				log.Infof("hole punching with %s successful after %s", rp, dt)
+				return nil
+			}
+		case <-hs.ctx.Done():
+			timer.Stop()
+			return hs.ctx.Err()
+		}
+	}
+	return fmt.Errorf("all retries for hole punch with peer %s failed", rp)
+}
+
+func (hs *Service) handlerError(p peer.ID, err error) {
+	hs.tracer.ProtocolError(p, err)
+	log.Warn(err)
+}
+
+func (hs *Service) handleNewStream(s network.Stream) {
+	log.Infof("got hole punch request from peer %s", s.Conn().RemotePeer().Pretty())
+	_ = s.SetDeadline(time.Now().Add(StreamTimeout))
+	rp := s.Conn().RemotePeer()
+	wr := protoio.NewDelimitedWriter(s)
+	rd := protoio.NewDelimitedReader(s, maxMsgSize)
+
+	// Read Connect message
+	msg := new(pb.HolePunch)
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("failed to read message from initator: %w", err))
+		return
+	}
+	if t := msg.GetType(); t != pb.HolePunch_CONNECT {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("expected CONNECT message from initiator but got %d", t))
+		return
+	}
+	obsDial := addrsFromBytes(msg.ObsAddrs)
+
+	// Write CONNECT message
+	msg.Reset()
+	msg.Type = pb.HolePunch_CONNECT.Enum()
+	msg.ObsAddrs = addrsToBytes(hs.ids.OwnObservedAddrs())
+	tstart := time.Now()
+	if err := wr.WriteMsg(msg); err != nil {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("failed to write CONNECT message to initator:: %w", err))
+		return
+	}
+
+	// Read SYNC message
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("failed to read message from initator: %w", err))
```

Comment:
Can we defer these? Even better, use a helper function and reset & log the error if the helper function fails.

Reply:
Yes, good point. I didn't like `handlerError` to begin with.